remote/bitbucketserver/internal/client.go

```diff
@@ -0,0 +1,234 @@
+package internal
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/drone/drone/model"
+	"github.com/mrjones/oauth"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const (
+	currentUserId   = "%s/plugins/servlet/applinks/whoami"
+	pathUser        = "%s/rest/api/1.0/users/%s"
+	pathRepo        = "%s/rest/api/1.0/projects/%s/repos/%s"
+	pathRepos       = "%s/rest/api/1.0/repos?limit=%s"
+	pathHook        = "%s/rest/api/1.0/projects/%s/repos/%s/settings/hooks/%s"
+	pathSource      = "%s/projects/%s/repos/%s/browse/%s?raw"
+	hookName        = "com.atlassian.stash.plugin.stash-web-post-receive-hooks-plugin:postReceiveHook"
+	pathHookEnabled = "%s/rest/api/1.0/projects/%s/repos/%s/settings/hooks/%s/enabled"
+)
+
+type Client struct {
+	client      *http.Client
+	base        string
+	accessToken string
+}
+
+func NewClientWithToken(url string, Consumer *oauth.Consumer, AccessToken string) *Client {
+	var token oauth.AccessToken
+	token.Token = AccessToken
+	client, err := Consumer.MakeHttpClient(&token)
+	log.Debug(fmt.Printf("Create client: %+v %s\n", token, url))
+	if err != nil {
+		log.Error(err)
+	}
+	return &Client{client, url, AccessToken}
+}
+
+func (c *Client) FindCurrentUser() (*model.User, error) {
+	CurrentUserIdResponse, err := c.client.Get(fmt.Sprintf(currentUserId, c.base))
+	if err != nil {
+		return nil, err
+	}
+	defer CurrentUserIdResponse.Body.Close()
+	bits, err := ioutil.ReadAll(CurrentUserIdResponse.Body)
+	if err != nil {
+		return nil, err
+	}
+	login := string(bits)
+
+	CurrentUserResponse, err := c.client.Get(fmt.Sprintf(pathUser, c.base, login))
+	if err != nil {
+		return nil, err
+	}
+	defer CurrentUserResponse.Body.Close()
+
+	contents, err := ioutil.ReadAll(CurrentUserResponse.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var user User
+	err = json.Unmarshal(contents, &user)
+	if err != nil {
+		return nil, err
+	}
+
+	ModelUser := &model.User{
```

Comment:
also if you look at the bitbucket implementation, the bitbucket client returns native bitbucket types. These types are then converted in a `convert.go` file. This allows use to change the drone model without having to change the client. Also much easier to unit test.
https://github.com/drone/drone/blob/master/remote/bitbucket/convert.go